<a href="https://colab.research.google.com/github/matherialist/dota-2-win-probability-prediction/blob/master/DOTA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import datetime

In [0]:
features = pd.read_csv('https://d3c33hcgiwev3.cloudfront.net/_47ea8666d262362a0b41164cb497658e_features.zip?Expires=1576195200&Signature=JsDvxiIFoNW7EpdBeZ2pEVrr3KHMETtQlmAuuD-~za0YRU83JlSjSZ6fTeP6pHif~8Iv5KQC0ieVZboGmmC1xRvwHie2JafzoNtMsZl-X~SuWdaZ6hN~z~nbbquhwHYSteUUaaHvo3mdYbodhlbk2zqx3CapEDFIPebqz0azVBY_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A', compression='zip', index_col='match_id')

In [3]:
features.head()

start_time  lobby_type  ...  barracks_status_radiant  barracks_status_dire
match_id                          ...                                               
0         1430198770           7  ...                       51                     0
1         1430220345           0  ...                       63                     1
2         1430227081           7  ...                        0                    63
3         1430263531           1  ...                       50                    63
4         1430282290           7  ...                        3                    63

[5 rows x 108 columns]

In [0]:
y_train = features['radiant_win']
X_train = features.loc[:, features.columns != 'radiant_win']

In [0]:
features_test = pd.read_csv('https://d3c33hcgiwev3.cloudfront.net/_6866d458468373e164b1e859f3103826_features_test.zip?Expires=1576195200&Signature=AkH0bulkEwmdyJFfiXfsKYX2Zaod23QaH-DNg9cgKsagKbilTeTFHBX4~fYKbeZCW3MJQrlKmnuGpoOMYITRUk9nFwhNCFaQbclzq6iUvSSv4ymO5AwOGvOagOgxfRTnRbYQUNPT~vTiOp~BbWJ3ypGvFW6bB2DJO3k2v0MfbhA_&Key-Pair-Id=APKAJLTNE6QMUY6HBC5A', compression='zip', index_col='match_id')

In [0]:
features_test.head()

start_time  lobby_type  ...  dire_ward_sentry_count  dire_first_ward_time
match_id                          ...                                              
6         1430287923           0  ...                       0                 118.0
7         1430293357           1  ...                       1                  16.0
10        1430301774           1  ...                       0                 -34.0
13        1430323933           1  ...                       1                 -26.0
16        1430331112           1  ...                       1                 -33.0

[5 rows x 102 columns]

**Подход 1: градиентный бустинг "в лоб"**

In [0]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import KFold, cross_val_score

In [0]:
X_train = features.drop(['duration', 'radiant_win', 'tower_status_radiant', 
                            'tower_status_dire', 'barracks_status_radiant', 
                            'barracks_status_dire'], axis=1)

In [0]:
X_train.shape

(97230, 102)

Определим признаки, содержащие пропуски.

In [0]:
X_train.columns[X_train.isna().any()].tolist()

['first_blood_time',
 'first_blood_team',
 'first_blood_player1',
 'first_blood_player2',
 'radiant_bottle_time',
 'radiant_courier_time',
 'radiant_flying_courier_time',
 'radiant_first_ward_time',
 'dire_bottle_time',
 'dire_courier_time',
 'dire_flying_courier_time',
 'dire_first_ward_time']

Если событие "первая кровь" не успело произойти за первые 5 минут, то признаки first_blood_time, first_blood_team, first_blood_player1, first_blood_player2 принимают пропущенное значение

In [0]:
X_train.fillna(0, inplace=True)

In [0]:
X_train.columns[X_train.isna().any()].tolist()

[]

Целевая переменная - radiant_win

In [0]:
y_train = features['radiant_win']

In [0]:
kf = KFold(n_splits=5, shuffle=True)

for n in [10, 20 , 30, 40, 50, 100]:
    clf = GradientBoostingClassifier(n_estimators=n, verbose=True)
    start_time = datetime.datetime.now()
    score = cross_val_score(clf, X_train, y_train, cv=kf, scoring="roc_auc", n_jobs=-1).mean()

    print('n_estimators =', n)
    print('Score:', score)
    print('Time elapsed:', datetime.datetime.now() - start_time)
    print()

n_estimators = 10
Score: 0.6644723421723491
Time elapsed: 0:00:25.247362

n_estimators = 20
Score: 0.6820569226035162
Time elapsed: 0:00:45.669416

n_estimators = 30
Score: 0.6898865269074778
Time elapsed: 0:01:05.856615

n_estimators = 40
Score: 0.6942058064933108
Time elapsed: 0:01:26.839650

n_estimators = 50
Score: 0.697810155255931
Time elapsed: 0:01:47.103974

n_estimators = 100
Score: 0.7068091429057917
Time elapsed: 0:03:31.404442



Для градиентного бустинга с 30 деревьями кросс-валидация длилась 0:01:05.856615, значение ROC AUC при этом 0.6899. 
Есть смысл увеличивать число деревьев, т.к. это ведёт к увеличению метрики ROC AUC.

**Подход 2: логистическая регрессия**

In [0]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler

In [0]:
X_train = features.drop(['duration', 'radiant_win', 'tower_status_radiant', 
                            'tower_status_dire', 'barracks_status_radiant', 
                            'barracks_status_dire'], axis=1)
X_train.fillna(0, inplace=True)

In [0]:
y_train = features['radiant_win']

In [0]:
scaler = StandardScaler()
X_train = pd.DataFrame(scaler.fit_transform(X_train), index=X_train.index, columns=X_train.columns)

In [20]:
kf = KFold(n_splits=5, shuffle=True)

for i in range(-5, 5):
    C = 10.0 ** i
    clf = LogisticRegression(penalty='l2', C=C, verbose=True)
    start_time = datetime.datetime.now()
    score = cross_val_score(clf, X_train, y_train, cv=kf, scoring="roc_auc", n_jobs=-1).mean()

    print('C =', C)
    print('Score:', score)
    print('Time elapsed:', datetime.datetime.now() - start_time)
    print()

C = 1e-05
Score: 0.6951203930965234
Time elapsed: 0:00:04.398966

C = 0.0001
Score: 0.7113197916153771
Time elapsed: 0:00:05.981815

C = 0.001
Score: 0.7163510937876062
Time elapsed: 0:00:09.334101

C = 0.01
Score: 0.7163585184650173
Time elapsed: 0:00:12.308196

C = 0.1
Score: 0.7165891758956429
Time elapsed: 0:00:13.276702

C = 1.0
Score: 0.7163025288197954
Time elapsed: 0:00:12.894827

C = 10.0
Score: 0.7164020195941407
Time elapsed: 0:00:13.112117

C = 100.0
Score: 0.7163863410507656
Time elapsed: 0:00:12.884006

C = 1000.0
Score: 0.7164502687881228
Time elapsed: 0:00:12.942883

C = 10000.0
Score: 0.7163485320147955
Time elapsed: 0:00:12.987991

